java/daikon/dcomp/DCInstrument.java

```diff
@@ -1928,26 +2015,87 @@ InstructionList handle_invoke(InvokeInstruction invoke) {
+            && !is_object_equals(method_name, ret_type, arg_types)) {
+
           // debug code
-          // System.out.printf("invoke virtual: %s : %s%n", classname, method_name);
-          // System.out.printf("super class: %s%n", gen.getSuperclassName());
-          // Technically, we should verify the target class has super class
-          // of java.lang.Enum. But that can be difficult if we haven't already
-          // processed that class. And since the worst that happens is we
-          // loose some tag interactions, we just go ahead.
-          if (method_name.equals("ordinal")) {
-            callee_instrumented = false;
-          }
+          // System.out.println("method: " + method_name);
+          // System.out.println("arg_types: " + Arrays.toString(arg_types));
+          // System.out.printf("invoke host: %s%n", gen.getClassName() + "." + mgen.getName());
+
+          @ClassGetName String targetClassname = classname;
+          mainloop:
+          while (true) {
+            // Check that the class exists
+            JavaClass targetClass;
+            try {
+              targetClass = getJavaClass(targetClassname);
+            } catch (Throwable e) {
+              targetClass = null;
+            }
+            if (targetClass == null) {
+              // We cannot locate or read the .class file, better assume not instrumented.
+              // debug code
+              System.out.printf("Unable to locate class: %s%n%n", targetClassname);
+              callee_instrumented = false;
+              break;
+            }
+            // debug code
+            // System.out.println("target class: " + targetClassname);
+
+            for (Method m : targetClass.getMethods()) {
+              // debug code
+              // System.out.println("  " + m.getName() + Arrays.toString(m.getArgumentTypes()));
+              if (m.getName().equals(method_name)
+                  && Arrays.equals(m.getArgumentTypes(), arg_types)) {
+                // We have a match.
+                // debug code
+                // System.out.printf("we have a match%n%n");
+                if (BcelUtil.inJdk(targetClassname)) {
+                  callee_instrumented = false;
+                }
+                break mainloop;
+              }
+            }
 
-          // TODO:
-          // This is actually a general problem.  Correct solution would seem
-          // to be finding the target of virtual call at run time.
-          // This is just a hack to get through PASCALI corpus.
-          String super_class = gen.getSuperclassName();
-          if (!super_class.equals("java.lang.Object") && BcelUtil.inJdk(super_class)) {
-            callee_instrumented = false;
+            // no methods match - does this class implement interfaces?
+            @ClassGetName String found;
+            try {
+              found =
+                  search_class_interfaces_for_target_method(targetClass, method_name, arg_types);
```

Comment:
Why does this method only search interfaces, rather than also superclasses (that is, all supertypes)?  Doing so could make the implementation cleaner here and eleminate the need for the `while (true)` loop (though the purpose of the code block should still be documented).

Reply:
Because we need to search all the interfaces of the current class before we recurse to its super class.